### ISSUE TITLE
Numerical instability in truncated normal initialization with bfloat16

### ISSUE THREAD

**Issue by francesco-bertolotti** [pytorch/torchtitan#2269]
### Bug description

I am experiencing numerical instability issues in training with qwen3. I tracked down the issue to the initialization of the model. In particular, to the usage of `torch.nn.init.trunc_normal_`. The issue is illustrated in the following code snippet. It is apparent that this function does not produce the intended behavior and it is skewed heavily toward the argument of `a`, in this case `-2`. I see that this function has an history of being problematic (see https://github.com/pytorch/pytorch/issues/155588 and https://github.com/pytorch/pytorch/issues/145498)

```python
In [20]: import torch
    ...: x = torch.zeros(10,500, dtype=torch.bfloat16)
    ...: torch.nn.init.trunc_normal_(x, mean=0, std=0.02, a=-2, b=2)
    ...: print("min", x.min(-1)[0])
    ...: print("max", x.max(-1)[0])
    ...: print("mean", x.mean(-1)[0])
    ...: print("sum", x.sum(-1)[0])
```
```
min tensor([-2.0000, -2.0000, -2.0000, -2.0000, -2.0000, -0.0532, -2.0000, -2.0000, -2.0000, -2.0000], dtype=torch.bfloat16)
max tensor([0.0532, 0.0532, 0.0532, 0.0532, 0.0532, 0.0532, 0.0532, 0.0532, 0.0532, 0.0483], dtype=torch.bfloat16)
mean tensor(-0.0084, dtype=torch.bfloat16)
sum tensor(-4.1875, dtype=torch.bfloat16)
```

My current workaround is to cast to float32, use truncated_init_, and cast back to bfloat16. This is the related code-snippet

```python
@torch.no_grad()
def trunc_normal_(tensor, mean=0.0, std=1.0, a=-2.0, b=2.0):
    orig_dtype = tensor.dtype

    if isinstance(tensor, torch.distributed.tensor.DTensor):
        # Convert DTensor -> float32 DTensor
        tensor_fp32 = tensor.to(dtype=torch.float32)

        # Local init (safe because trunc_normal_ is elementwise)
        torch.nn.init.trunc_normal_(tensor_fp32._local_tensor, mean=mean, std=std, a=a, b=b)

        # Cast back to original dtype and copy into original tensor
        tensor.copy_(tensor_fp32.to(dtype=orig_dtype))
    else:
        # Regular Tensor
        tensor_fp32 = tensor.to(dtype=torch.float32)
        torch.nn.init.trunc_normal_(tensor_fp32, mean=mean, std=std, a=a, b=b)
        tensor.copy_(tensor_fp32.to(dtype=orig_dtype))

    del tensor_fp32

    return tensor
```

----

Please, let me know if you have any suggestion regarding this issue. I can also submit a PR with the proposed workaround.

### Versions

torch: 2.11.0.dev20260107+cu126

**Comment by tianyu-l** [comment]
@albanD what's your latest take on this issue? Do you think this is not a bug, but expected with bf16?

**Comment by tianyu-l** [comment]
@francesco-bertolotti Curious: are you training qwen3 from scratch, instead of loading a checkpoint? Would love to hear more about your use case.

**Comment by francesco-bertolotti** [comment]
Hi @tianyu-l,

Yes, we’re essentially iterating on the qwen3 architecture and pretraining from scratch. I’m trying to ensure the training process runs as smoothly as possible, so I’m also taking a close look at the activations.

**Comment by francesco-bertolotti** [comment]
Small followup. 

I found that the suggested workaround was affected by the same problem. I find this implementation to be more stable, though it could loop indefinitely when `a` is close to `b`

```python

@torch.no_grad()
def trunc_normal_(tensor, mean=0.0, std=1.0, a=-2.0, b=2.0):
    if isinstance(tensor, torch.distributed.tensor.DTensor):
        tensor = tensor.to_local()

    tensor.normal_(mean, std)
    while tensor.lt(a).any() or tensor.gt(b).any():
        tensor[tensor.lt(a).logical_or(tensor.gt(b))] = tensor[tensor.lt(a).logical_or(tensor.gt(b))].normal_(mean, std)
    
    return tensor
```

**Comment by francesco-bertolotti** [comment]
Hi @tianyu-l, I see this has been confirmed as a bug. If the last proposed workaround is an acceptable fix, I’m happy to put together a PR for it. Let me know if you’d like me to proceed!

**Comment by albanD** [comment]
> I found that the suggested workaround was affected by the same problem. 

Which workaround? Sampling in fp32 then down-casting?

> what's your latest take on this issue? Do you think this is not a bug, but expected with bf16?

It is a pretty subtle one for sure.
Yes you should expect the distribution to be worse when sampling in bf16.
But I also agree that it shouldn't be that bad!
The question is where do we want to put the tradeoff between extra compute and quality of the result.

For torchtitan, if you know you always want good quality sampling. I think upscaling to fp32 to do the sampling is a good fix.
For core, we would need to study the sampling function more closely to try and generate a better distribution given the limited randomness and without a significant slowdown (if this is memory bound, doing fp32 in-kernel might be enough). Someone would have to prepare such a PR and carefully validate the accuracy of it.

**Comment by francesco-bertolotti** [comment]
> Which workaround? Sampling in fp32 and then down-casting?

Yes. Using `init.trunc_normal_` in fp32 and casting to bf16 exhibits the same issue---specifically, anomalous sampling at the left bound.

> The question is where we want to place the tradeoff between extra compute and result quality.

For PyTorch, it’s probably preferable to provide a fast implementation that is as accurate as possible, with any lower-precision instabilities documented. For TorchTitan, however, I think it makes sense to trade some performance for a more stable initialization.

My reasoning is that this issue surfaced when early-training activations in a Qwen3 model became unusually large. Replacing the unstable function with a stable one immediately reduced activation magnitudes and lowered the loss. Additionally, with standard initialization parameters, we only observed 2-3 resampling events, and only for the largest tensors (the embeddings).

I’ll follow up in the next few days with more thorough documentation of the issue, potentially including plots.

**Comment by albanD** [comment]
@francesco-bertolotti from this small test script, fp32 sampling downcasted does work well.
It feels like we are in agreement that in the context of torchtitan, fp32 sampling that is then downcasted is the right thing to do for now cc @tianyu-l 


Details on the sampling

<img width="1194" height="410" alt="Image" src="https://github.com/user-attachments/assets/668b999d-1908-4451-ba34-ae16cfdbaa03" />

```python
import torch
import matplotlib.pyplot as plt

n = 100000
kwargs = dict(mean=0, std=0.02, a=-2, b=2)

fp32 = torch.nn.init.trunc_normal_(torch.empty(n), **kwargs)
fp32_to_bf16 = torch.nn.init.trunc_normal_(torch.empty(n), **kwargs).bfloat16()
bf16 = torch.nn.init.trunc_normal_(torch.empty(n, dtype=torch.bfloat16), **kwargs)

samples = [("fp32", fp32), ("fp32→bf16", fp32_to_bf16), ("bf16", bf16)]
for name, data in samples:
    print(f"{name}: {len(data.unique())} unique values")

fig, axes = plt.subplots(1, 3, figsize=(12, 4), sharey=True)
for ax, (name, data) in zip(axes, samples):
    ax.hist(data.float().numpy(), bins=100)
    ax.axvline(data.float().mean(), color="red", linestyle="--", label=f"mean={data.float().mean():.4f}")
    ax.set_title(f"{name}\nmin={data.min():.3f} max={data.max():.3f}")
    ax.set_xlim(-0.1, 0.1)
    ax.legend()

plt.tight_layout()
plt.savefig("trunc_normal_comparison.png", dpi=150)
plt.show()

```

Output:
```
fp32: 99728 unique values
fp32→bf16: 2681 unique values
bf16: 256 unique values
```


The number of unique values that were sampled might hint at a problem in the amount of randomness used when sampling in bf16 natively which could explain this ? Would need more investigation

**Comment by francesco-bertolotti** [comment]
Interesting, there must have been something not working in my first implementation because I was still seeing instabilities. This weekend I should be able to submit a first a PR for this. `torchtitan/models/utils.py` looks like a good place for it

**Comment by tianyu-l** [comment]
Maybe should re-test pytorch trunk_normal after https://github.com/pytorch/pytorch/pull/174997 lands

**Comment by tianyu-l** [comment]
@francesco-bertolotti since https://github.com/pytorch/pytorch/pull/174997 got merged, would you be interested in testing it out and maybe change back to use the "official" solution?

turns out you also need https://github.com/pytorch/pytorch/pull/176063

**Comment by francesco-bertolotti** [comment]
Yes, I should be able to find some time next week.

**Comment by albanD** [comment]
This should be fixed on trunk now!
Also if you're doing meta device initialization, make sure to have https://github.com/pytorch/pytorch/pull/176240 !

**Comment by wwwjn** [comment]
@francesco-bertolotti  Can you check if it fix the issue?  If so, we can mark this issue closed

**Comment by francesco-bertolotti** [comment]
Yep, i am planning to do it. I had trouble finding the time this week. I should be able to do a few tests and to submit a PR in the coming days.
